relay-quotas/src/redis.rs

```diff
@@ -125,6 +139,300 @@ impl std::ops::Deref for RedisQuota<'_> {
+#[derive(Debug, thiserror::Error)]
+enum GlobalRateLimitError {
+    #[error("quota key missing")]
+    KeyMissing,
+    #[error("failed to access RWLock")]
+    PoisonedLock,
+    #[error("slot expired")]
+    SlotExpired,
+    #[error("budget empty")]
+    BudgetEmpty,
+    #[error("failed to communicate with redis")]
+    Redis,
+    #[error("stuck in loop")]
+    LoopLimitExceeded,
+}
+
+/// Counters used as a cache for global quotas.
+///
+/// When we want to ratelimit across all relay-instances, we need to use redis to synchronize.
+/// Calling Redis every time we want to check if an item should be ratelimited would be very expensive,
+/// which is why we have this cache. It works by 'taking' a certain budget from redis, by pre-incrementing
+/// a global counter. We Put the amount we pre-incremented into this local cache and count down until
+/// we have no more budget, then we ask for more from redis. If we find the global counter is above
+/// the quota limit, we will ratelimit the item.
+#[derive(Clone)]
+struct GlobalCounters {
+    script: Arc<Script>,
+    counters: Arc<RwLock<hashbrown::HashMap<BudgetKey, BudgetState>>>,
+}
+
+impl Default for GlobalCounters {
+    fn default() -> Self {
+        Self {
+            script: Arc::new(load_global_lua_script()),
+            counters: Default::default(),
+        }
+    }
+}
+
+impl fmt::Debug for GlobalCounters {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut debug_struct = f.debug_struct("GlobalCounters");
+
+        if let Ok(counters) = self.counters.read() {
+            for (key, val) in counters.iter() {
+                let budget = val.budget.load(Ordering::SeqCst);
+                let slot = val.slot.load(Ordering::SeqCst);
+                let redis = val.last_seen_redis_value.load(Ordering::SeqCst);
+
+                debug_struct.field(
+                    &format!("{:?}", key),
+                    &format!("budget: {}, redis_value: {}, slot {}", budget, redis, slot),
+                );
+            }
+        } else {
+            debug_struct.field("counters", &"Lock Poisoned");
+        }
+
+        debug_struct.finish()
+    }
+}
+
+impl GlobalCounters {
+    /// Returns `true` if the global quota should be ratelimited.
+    ///
+    /// Certain errors can be resolved by syncing to redis, so in those cases
+    /// we try again to decrement the budget after syncing.
+    fn evaluate_quota(
+        &self,
+        client: &mut PooledClient,
+        quota: &RedisQuota,
+        quantity: usize,
+    ) -> Result<bool, GlobalRateLimitError> {
+        use GlobalRateLimitError as E;
+
+        match self.decrement_budget(quota, quantity) {
+            ok @ Ok(_) => ok,
+            err @ Err(E::Redis | E::PoisonedLock | E::LoopLimitExceeded) => err,
+            Err(E::BudgetEmpty | E::SlotExpired | E::KeyMissing) => {
+                self.redis_sync(client, quota, quantity)?;
+                self.decrement_budget(quota, quantity)
```

Comment:
the slot could technically have expired between the two calls, but the chance of that is too small to bother handling imo